docker-compose.yaml

```diff
@@ -6,18 +6,16 @@ services:
             context: https://github.com/Chemical-Curation/chemcurator_django.git#${API_BRANCH}
         image: chemreg-django
         restart: unless-stopped
+        env_file:
+          - .env
         environment:
             CACHE_URL: redis://redis:6379/0
```

Comment:
environment: takes priority over env_file:
https://docs.docker.com/compose/environment-variables/#the-env-file
so some of these are hard coded here and not configurable without editing docker-compose.yml.

Reply:
The hardcoded `5432` could be changed to `${SQL_PORT}`, I agree. Which other hardcoded variables do you think are worth parameterizing into the `.env`?

Reply:
Not so much a question of individual settings but the flow of env. vars. in general.
- This PR does what the ticket requested, so maybe these issues should be a different ticket, not sure how much work you typically do in the PR phase and how that interacts with pointing etc.
- I could be wrong about any of this, Docker's variable handling is a mess.
- I'm assuming that ".env" comes from https://github.com/Chemical-Curation/chemcurator_django/blob/d4545cc/template.env.
- Many of the variables in the template.env / .env are listed in the docker-compose environment: section.  It's my understanding that the latter has priority, so listing them in .env has no effect, but possibly creates confusion if people try and change them in .env.
- If https://github.com/Chemical-Curation/chemcurator_django/blob/17410f1/config/settings.py#L24 does what it appears to do, I get the impressing the Django process is bypassing the Docker variable handling, paying attention to the .env file again, but creating the potential for different processes to see different values for "environment" variables.
- At this point it seems clear this should be a separate ticket :-)

Reply:
Yes, whenever you've identified either a bug or new functionality that is outside the immediate scope of the ticket, making a new ticket is in order. Then it can be added to the backlog and the product owner can evaluate when to include it in a sprint.

Reply:
@tbnorth We've used different approaches to env management.  The way we do it on our related project, [factotum](https://github.com/HumanExposure/factotum/blob/master/docker/docker-compose.override.yaml), is by listing every environment variable explicitly in the container and not use the .env at all outside of that.  This made it so env variables would not only need to be added to the code, but to the compose file as well.

This had the side effect where we would build a new variable into the code then push it up and require a hotfix to add it to the compose.  To get around this I'd like us to attempt to leverage the env hierarchy.   

Variables **explicitly added to the compose file** are built to be overrides.  These are things that are explicitly implied by the container you're building.  For instance if you're building `chemreg-api`, the URLs we should be using should be api's urls.  Thus no matter what you have in your `.env`, the [URL_CONF](https://github.com/Chemical-Curation/chemcurator_django/wiki/Configuration#url_conf-toc) for this container should be `URL_CONF: api`.

The **env_file** variables are expected to be variables that are purely for the app and not reliant on which container you're attempting to build.  For instance the [RESOLUTION_URL](https://github.com/Chemical-Curation/chemcurator_django/wiki/Configuration#resolution_url-toc) will probably not change between the admin or api containers.  

This setup is not perfect.  As @mwfrost said, we will need to customize the ports (similar to resolver [#9](https://github.com/Chemical-Curation/resolver/issues/9)), and I think we should consider some of the other container overrides.  But first I was hoping to get this as similar to prod as possible before we make those changes.

### A quick rundown of the hardcoded env variables and their reasons:

- CACHE_URL: Hardcoded to use the redis container that developers may not have.
- DATABASE_URL: Hardcoded to use the pgbouncer container.  This was added in an attempt to speed up pgsql connections
- SECRET_KEY: Each container should have their own env.
- URL_CONF: The primary difference between containers.  Decided which endpoints will be served
- WEB_CONCURRENCY: Honestly, I'm not sure.  It was that way on the [docker-compose.yaml](https://github.com/Chemical-Curation/chemcurator_django/wiki/Deploying-with-Docker#docker-composeyaml) that we had been using.  I'm assuming it's to give the api sever more ability to respond to requests, but mostly I feel uncomfortable changing it without knowledge.
- WHITELIST_CORS: This should probably be set to `None` for admin.  Allows our SPA to communicate with our backend.  [WHITELIST_LOCAL](https://github.com/Chemical-Curation/chemcurator_django/wiki/Configuration#whitelist_local-TOC) still works and should be None for prod, but that is an env variable and one that developers will need to test the docker-compose setup.
- WHITELIST_HOST: The url that the container lives at.  As both containers may have different routing/urls, this needs to be overridden with a container specific value.

Reply:
Thanks for the background.  So I wonder if `URL_CONF` should be left out of `template.env`, if it can't be edited there? I realize there may be factors I'm missing here.
Also it looks to me as if https://github.com/Chemical-Curation/chemcurator_django/blob/17410f1/config/settings.py#L24 will cause Django to ignore the `environment:` section in `docker-compose.yml`, although the fact that it's working so far kind of undermines that assertion :)

I'm trying to put together a efficient deployment path to deploy ChemReg on shared host for testing / PR review.  I.e. the system that we're being told to use for testing / PR review could have two reviewers wanting to spin up different branches of ChemReg on the same host.  If this results in endless torment the shared host idea may need to be revisited.  I had got as far as reading the node/vue `docker-compose.yml` and programmatically adjusting the exposed ports (e.g. 37000,37001,37002,37003 in place of 8000,5000,5432,8081 or whatever they are).  My code programmatically added services for the django containers.  That was before this PR, so next go around will read and merge both `docker-compose.yml` files before tweaking things.  Kind of wonder if docker-in-docker would be a simpler path.  Anyway, apart from general interst in Docker, that's part of the source of my interest.

Reply:
> Thanks for the background. So I wonder if `URL_CONF` should be left out of `template.env`, if it can't be edited there? I realize there may be factors I'm missing here.

It's mostly there for developers to be able to spin it up.  If a dev needed to check the admin app, then the `URL_CONF` in the `.env` would be the approach.  I'm in favor of finding a good balance for our settings but unless this is actively confusing the deployment, I'd like all our `.env` variables to be in the template. (nobody like's combing docs for hidden variables)

> I'm trying to put together a efficient deployment path to deploy ChemReg on shared host for testing / PR review. I.e. the system that we're being told to use for testing / PR review could have two reviewers wanting to spin up different branches of ChemReg on the same host. If this results in endless torment the shared host idea may need to be revisited. I had got as far as reading the node/vue `docker-compose.yml` and programmatically adjusting the exposed ports (e.g. 37000,37001,37002,37003 in place of 8000,5000,5432,8081 or whatever they are). My code programmatically added services for the django containers. That was before this PR, so next go around will read and merge both `docker-compose.yml` files before tweaking things. Kind of wonder if docker-in-docker would be a simpler path. Anyway, apart from general interst in Docker, that's part of the source of my interest.

I am currently working with development and a very simple production build in mind.  There are a lot of wonderful solutions to CI/CD out there, but for now it seems our attention is being focused on building the product.  If a deployment strategy change is requested, I'd be happy to accommodate or provide potential solutions, but for now this approach seems to be working.